File: oop/rich/src/main/java/de/hhu/rich/warenkorb/domain/model/artikel/Artikel.java
package de.hhu.rich.warenkorb.domain.model.artikel;

import de.hhu.rich.warenkorb.domain.model.PflichtfeldFehlt;
import de.hhu.rich.warenkorb.domain.model.common.preis.Preis;

import java.math.BigDecimal;

public class Artikel {
    private final ArtikelID ID;
    private String name;
    private String beschreibung;
    private Preis preis;

    public Artikel(ArtikelID ID, String name, String beschreibung, Preis preis) {
        this.ID = ID;
        this.name = name;
        this.beschreibung = beschreibung;
        this.preis = preis;
        validiere();
    }

    public ArtikelID getID() {
        return ID;
    }

    public String getName() {
        return name;
    }

    public String getBeschreibung() {
        return beschreibung;
    }

    public Preis getPreis() {
        return preis;
    }

    private void validiere() {
        if(ID == null) {
            throw new PflichtfeldFehlt("ArtikelID");
        }
        if (name == null || name.trim().isEmpty()) {
            throw new PflichtfeldFehlt("Artikel Name");
        }
        if(beschreibung == null || beschreibung.trim().isEmpty()) {
            throw new PflichtfeldFehlt("Artikel Beschreibung");
        }
        if (preis == null) {
            throw new PflichtfeldFehlt("Artikel Preis");
        }
    }

    public void setzeNeuenNamen(String neuerName) {
        if (neuerName == null || neuerName.trim().isEmpty()) {
            throw new PflichtfeldFehlt("Neuer Name darf nicht null oder leer sein.");
        }
        this.name = neuerName;
    }

    public void setzeNeuenPreis(Preis neuerPreis) {
        if (neuerPreis == null || neuerPreis.betrag().compareTo(BigDecimal.ZERO) < 0) {
            throw new PflichtfeldFehlt("Neuer Preis darf nicht null sein und muss nicht-negativ sein.");
        }
        this.preis = neuerPreis;
    }

    public void setzeNeueBeschreibung(String neueBeschreibung) {
        if (neueBeschreibung == null || neueBeschreibung.trim().isEmpty()) {
            throw new PflichtfeldFehlt("Neue Beschreibung darf nicht null oder leer sein.");
        }
        this.beschreibung = neueBeschreibung;
    }

}

File: fp/rich/src/main/clojure/de/hhu/rich/warenkorb/domain/model/artikel/Artikel.clj
(ns de.hhu.rich.warenkorb.domain.model.artikel.Artikel
  (:require [de.hhu.rich.warenkorb.domain.model.artikel.ArtikelID :as artikelID]
            [de.hhu.rich.warenkorb.domain.model.preis.Preis :as preis]))

(defprotocol IArtikel
  (get-id [this])
  (get-name [this])
  (get-beschreibung [this])
  (get-preis [this])
  (set-name [this new-name])
  (set-beschreibung [this new-beschreibung])
  (set-preis [this new-preis]))

(defrecord Artikel [id name beschreibung preis]
  IArtikel
  (get-id [_] id)
  (get-name [_] name)
  (get-beschreibung [_] beschreibung)
  (get-preis [_] preis)
  (set-name [this new-name]
    (if (or (nil? new-name) (empty? (clojure.string/trim new-name)))
      (throw (IllegalArgumentException. "Neuer Name darf nicht null oder leer sein."))
      (assoc this :name new-name)))
  (set-beschreibung [this new-beschreibung]
    (if (or (nil? new-beschreibung) (empty? (clojure.string/trim new-beschreibung)))
      (throw (IllegalArgumentException. "Neue Beschreibung darf nicht null oder leer sein."))
      (assoc this :beschreibung new-beschreibung)))
  (set-preis [this new-preis]
    (if (or (nil? new-preis) (< (.compareTo (:betrag new-preis) BigDecimal/ZERO) 0))
      (throw (IllegalArgumentException. "Neuer Preis darf nicht null sein und muss nicht-negativ sein."))
      (assoc this :preis new-preis))))

(defn validiere-artikel [artikel]
  (when (or (nil? (:id artikel))
            (nil? (:name artikel)) (empty? (:name artikel))
            (nil? (:beschreibung artikel)) (empty? (:beschreibung artikel))
            (nil? (:preis artikel)))
    (throw (IllegalArgumentException. "Pflichtfelder fehlen im Artikel"))))

(defn erstelle-artikel [name beschreibung preis]
  (let [neuer-artikel (->Artikel (artikelID/erstelleArtikelID)
                                 name
                                 beschreibung
                                 (preis/erstelle-preis preis "EUR"))]
    (validiere-artikel neuer-artikel)
    neuer-artikel))
